### PR TITLE
[patch] Fix state loosing case

### LINKED ISSUE #4: avoid re-render onPrint
Bug: lost every component's state on print 

TODO
- Render tmp view into separate root.
- Place tmp root as fixed 0:0
- Don't touch main react root.

### PR DESCRIPTION
**Pull request by Serjobas**
closes #4

## Added aliases for props 
With new prop names API became more clear and more obvious

Now you should use
- printOnly ( for exclusive prop ) 
- single ( for main prop )

## Now print mode doesn't remove the state of your React app
Now when you proceed to print mode ( cmd/ctrl + P ), Print components are being rendered in the new DOM element, which disappears after you exit from print mode. Because of that your React app keeps it's state. 

##  Refactoring
index.js single file splitted into independent react components, some unused code was removed from the lib

## ts typings 
- [x] update or ask contributors do it

### CODE REVIEW

File: src/PrintProvider.jsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+    const { isInPrintPreview, printableNodes, hasSingle } = this.state;
+    const loose = this.props.loose || this.props.invert;
+
+    if (isInPrintPreview && printableNodes.length && !loose && !hasSingle) {
+      debug('render printable only', printableNodes);
+      const children = React.Children.map(printableNodes, (child,key) => {
+        return React.cloneElement(child, { key });
+      });
+
+      this.createRender(<div>{children}</div>);
```
Comment (a-x-) [inline src/PrintProvider.jsx]: The offset (you used above) is good enough solution. We're don't need a React.Portal-like thing, for now, isn't it?

File: src/styles.css [hunk 2]
```diff
@@ -2,16 +2,32 @@
   display: none;
 }
 @media print {
+  html, body {
+    width:210mm;
+  }
```
Comment (a-x-) [inline src/styles.css]: what is it?

File: src/styles.css [hunk 3]
```diff
@@ -2,16 +2,32 @@
...
     visibility: hidden;
   }
   .print {
     visibility: visible; /* visible element even nested in the hidden one */
   }
   .print._main {
-    position: fixed;
-    left: 0;
-    top: 0;
+    visibility: visible !important;
```
Comment (a-x-) [inline src/styles.css]: why you're using important?

visibility: visible can override parent's visibility: hidden out of the box w/o important hacks

**Comment by a-x-** [comment]
Update readme.md, add deprecation notes.
Look at the reference diff: https://github.com/a-x-/react-easy-print/compare/WIP

File: src/PrintProvider.jsx (hunk 1 above)
Comment (Serjobas) [reply src/PrintProvider.jsx]: Offset thing is being used only for (single, main) Print component, we want to use (single, main)  prop when we don't want to lose the state of Print component's children and we only want to print only one particular component.

In other cases (when we want to print more than one Print component) the lib will create an extra DOM node and will render the Print components there

File: src/styles.css (hunk 2 above)
Comment (Serjobas) [reply src/styles.css]: 210 mm - A4 width
We explicitly tell it to prevent cases when content is overwhelmed

File: src/styles.css (hunk 3 above)
Comment (Serjobas) [reply src/styles.css]: Yes, we don't need important here

File: src/Print.jsx [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  componentWillUnmount() {
+    if (this.props.name) {
+      debug('remove printable', this.props.name);
+      this.context.printProvider && this.context.printProvider.unregPrintable(this.props.name);
+    }
+  }
+
+  componentDidlUpdate(){
+    const bodyRect = document.body.getBoundingClientRect();
+    const elem = document.getElementById('single');
```
[the remark this answers is not in the document]
Comment (Serjobas) [reply src/Print.jsx]: I had some troubles with using ref, cause of implicit re-rendering and re-updating
I'll try to revise and refactor this

File: src/Print.jsx [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    const elemRect = elem && elem.getBoundingClientRect();
+    const offsetCool = elemRect && (elemRect.top - bodyRect.top);
```
[the remark this answers is not in the document]
Comment (Serjobas) [reply src/Print.jsx]: Ok

File: src/Print.jsx [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+    this.offsetCool = offsetCool;
```
[the remark this answers is not in the document]
Comment (Serjobas) [reply src/Print.jsx]: OK

**Comment by Serjobas** [comment]
fixed PR issues in the latest commit

File: src/PrintProvider.jsx (hunk 1 above)
Comment (a-x-) [reply src/PrintProvider.jsx]: We can print one particular component only using offsets and css visibility hacks w/o re-rendering anymore, am I correct?

File: src/styles.css (hunk 2 above)
Comment (a-x-) [reply src/styles.css]: It should be optional 
PrintProvider width = ‘A4’ | Number | other css values

**Comment by a-x-** [comment]
I replied you in some folded out comments

File: src/PrintProvider.jsx (hunk 1 above)
Comment (Serjobas) [reply src/PrintProvider.jsx]: @a-x- yes, and if there are several of Print components, we're re-rendering them in the new DOM node, when print mode is on, then deleting this DOM node.

File: src/Print.jsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (Serjobas) [reply src/Print.jsx]: Yep, I implemented ref ``ref`` 
And removed element DOM manipulation

File: src/PrintProvider.jsx [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+        regPrintable: this.regPrintable.bind(this),
+        unregPrintable: this.unregPrintable.bind(this),
+      },
+    };
+  }
+
+  regPrintable(key, node, isSingle) {
+    const loose = this.props.loose || this.props.invert;
+    debug('reg printable', key, node);
+    if (this.printableRegistry[key] !== undefined || loose) return;
```
Comment (a-x-) [inline src/PrintProvider.jsx]: || isSingle

File: src/PrintProvider.jsx [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    setTimeout(() => this.setState({
+      printableNodes: this.state.printableNodes.concat(node),
+      hasSingle: this.state.hasSingle || isSingle
+    }), 0);
+    this.printableRegistry[key] = this.state.printableNodes.length;
+  }
+
+  unregPrintable(key) {
+    const loose = this.props.loose || this.props.invert;
+    if (this.printableRegistry[key] === undefined || this.state.isInPrintPreview || loose) return;
```
Comment (a-x-) [inline src/PrintProvider.jsx]: || isSingle

File: src/Print.jsx [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+      const isSingle = (this.props.main || this.props.single);
+      this.context.printProvider && this.context.printProvider.regPrintable(this.props.name, <Print {...this.props} />, isSingle);
+    }
+
+
+    if (this.props.main || this.props.single) {
+      const bodyRect = document.body.getBoundingClientRect();
+      const elem = this.printElement;
+      const elemRect = elem && elem.getBoundingClientRect();
+      const printOffset = elemRect && ( elemRect.top - bodyRect.top );
```
Comment (a-x-) [inline src/Print.jsx]: and left too

**Comment by a-x-** [comment]
Remove the **example 5** and add **single** prop to the **example 1**

**Comment by a-x-** [comment]
### Errata

**Pull request merged**